diff.go

```diff
@@ -0,0 +1,467 @@
+		result.NumUnits = d.diffInts(bundle.NumUnits, len(model.Units))
+	}
+
+	if result.Empty() {
+		return nil
+	}
+	return result
+}
+
+func (d *differ) diffMachines() map[string]*MachineDiff {
+	unseen := set.NewStrings()
+	for name := range d.config.Model.Machines {
+		unseen.Add(name)
+	}
+	// Go through the machines from the bundle, but keep track of
+	// which model machines we've seen.
+	results := make(map[string]*MachineDiff)
+	for bundleName, bundleMachine := range d.config.Bundle.Machines {
+		modelName := d.toModelMachineName(bundleName)
+		unseen.Remove(modelName)
+
+		if bundleMachine == nil {
+			// This is equivalent to an empty machine spec.
+			bundleMachine = &charm.MachineSpec{}
+		}
+		modelMachine, found := d.config.Model.Machines[modelName]
+		if !found {
+			results[modelName] = &MachineDiff{Missing: ModelSide}
+			continue
+		}
+		diff := &MachineDiff{
+			Series: d.diffStrings(
+				bundleMachine.Series, modelMachine.Series,
+			),
+		}
+
+		if d.config.IncludeAnnotations {
+			diff.Annotations = d.diffAnnotations(
+				bundleMachine.Annotations, modelMachine.Annotations,
+			)
+		}
+
+		if !diff.Empty() {
+			results[modelName] = diff
+		}
+	}
+
+	// Add missing bundle machines for any model machines that weren't
+	// seen.
+	for _, modelName := range unseen.Values() {
+		results[modelName] = &MachineDiff{Missing: BundleSide}
+	}
+
+	if len(results) == 0 {
+		return nil
+	}
+	return results
+}
+
+func (d *differ) toModelMachineName(bundleMachineName string) string {
+	result, found := d.config.Model.MachineMap[bundleMachineName]
+	if !found {
+		// We always assume use-existing-machines.
+		return bundleMachineName
+	}
+	return result
+}
+
+func (d *differ) diffRelations() *RelationsDiff {
+	bundleSet := make(map[Relation]bool)
+	for _, relation := range d.config.Bundle.Relations {
+		bundleSet[relationFromEndpoints(relation)] = true
+	}
+
+	modelSet := make(map[Relation]bool)
+	var modelExtra []Relation
+	for _, original := range d.config.Model.Relations {
+		relation := canonicalRelation(original)
+		modelSet[relation] = true
+		_, found := bundleSet[relation]
+		if !found {
+			modelExtra = append(modelExtra, relation)
+		}
+	}
+
+	var bundleExtra []Relation
+	for relation := range bundleSet {
+		_, found := modelSet[relation]
+		if !found {
+			bundleExtra = append(bundleExtra, relation)
+		}
+	}
+
+	if len(bundleExtra) == 0 && len(modelExtra) == 0 {
+		return nil
+	}
+
+	sort.Slice(bundleExtra, relationLess(bundleExtra))
+	sort.Slice(modelExtra, relationLess(modelExtra))
+	return &RelationsDiff{
+		BundleExtra: toRelationSlices(bundleExtra),
+		ModelExtra:  toRelationSlices(modelExtra),
+	}
+}
+
+func (d *differ) diffAnnotations(bundle, model map[string]string) map[string]StringDiff {
+	all := set.NewStrings()
+	for name := range bundle {
+		all.Add(name)
+	}
+	for name := range model {
+		all.Add(name)
+	}
+	result := make(map[string]StringDiff)
+	for _, name := range all.Values() {
+		bundleValue := bundle[name]
```

Comment:
I'm not sure - what do you think?

Reply:
We discussed it and decided we're ok with this. It's only for annotations, where there aren't really defaults (like there are for config); in that case the difference between not set and empty is pretty minimal. Options are handled differently because they are an OptionDiff with `interface{}` values, so they can be nil if the option isn't set.